source/client-side-encryption/client-side-encryption.rst

```diff
@@ -1332,6 +1367,67 @@ on sockets. See the SDAM spec description of `cooldownMS <../source/server-disco
 Because single threaded drivers may exceed ``serverSelectionTimeoutMS`` by the
 duration of the topology scan, ``connectTimeoutMS`` is also reduced.
 
+What's the deal with metadataClient, keyVaultClient, and the internal client?
+-----------------------------------------------------------------------------
+
+When automatically encrypting a command, the driver runs:
+- a ``listCollections`` command to determine if the target collection
+has a remote schema. This uses the ``metadataClient``.
+- a ``find`` against the key vault collection to fetch keys. This uses the
+``keyVaultClient``.
+
+**Why not reuse the parent MongoClient?**
+
+These operations MUST NOT reuse the same connection pool as the parent
+``MongoClient`` configured with automatic encryption to avoid possible deadlock
+situations.
+
+Drivers supporting a connection pool (see `CMAP specification
+</source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst>`_)
+support an option for limiting the connection pool size: ``maxPoolSize``.
+
+Drivers need to check out a connection before serializing the command. If the
+``listCollections`` or ``find`` command during automatic encryption uses the same
+connection pool as the parent MongoClient, the application is susceptible to
+deadlocks.
+
+Using the same connection pool causes automatic encryption to check out multiple
+connections from the pool when processing a single command. If maxPoolSize=1,
+this is an immediate deadlock. If maxPoolSize=2, and two threads check out the
+first connection, they will deadlock attempting to check out the second.
+
+**Why are keyVaultClient and metadataClient separate exposed options?**
+
+The ``keyVaultClient`` supports the use case where the key vault collection is
+stored on a MongoDB cluster separate from the data-bearing cluster.
+
+The ``metadataClient`` is only used for ``listCollections`` against the
+data-bearing cluster.
+
+``listCollections`` responses are cached by libmongocrypt for one minute.
+
+The use pattern of the ``metadataClient`` will likely greatly differ from
+the parent ``MongoClient``.
+
+This is an exposed option to provide a way for users to configure (e.g. set a
+lower ``minPoolSize`` to avoid additional connections).
+
+**Why is the metadataClient not needed if bypassAutoEncryption=true**
+
+Because automatic decryption does not require the JSON schema.
+``listCollections`` is not run during automatic encryption.
+
+**Can the keyVaultClient and the metadataClient be the same?**
+
+Technically yes, but not all drivers support this. In some drivers, the
+``keyVaultClient`` and ``metadataClient`` is configured by passing client
+options, instead of a client object.
+
+If a ``metadataClient`` is passed, but a ``keyVaultClient`` is set, it would be
```

Comment:
I don't understand this paragraph.

Reply:
The point I want to make is that we could optimize out an extra client, but I'm proposing not to.

In particular, if you set a `metadataClient`, but do not set a `keyVaultClient`, we _could_ internally reuse the `metadataClient` to serve as the `keyVaultClient` (instead of creating an internal client for `keyVaultClient`).

It's a caveat I'm willing to hash out if there is disagreement. But I tried clarifying this paragraph.